### PR TITLE
directive: number range selection directive addition

### PR DESCRIPTION
**Pull request by nikofil**
- NEW Adds a directive for filtering according to a range of numbers.

### CODE REVIEW

File: src/invenio-search-js/invenioSearch.module.js [hunk 1]
```diff
@@ -356,9 +356,8 @@
     );
   }
 
-  invenioSearchController.$inject = [
-    '$scope', 'invenioSearchHandler', 'invenioSearchAPI'
-  ];
+  invenioSearchController.$inject = ['$scope', 'invenioSearchHandler',
+    'invenioSearchAPI'];
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js]: Can you please undo? :)

File: src/invenio-search-js/invenioSearch.module.js [hunk 2]
```diff
@@ -1338,6 +1337,135 @@
...
+        barcolor: '#2c3e50',
+        selectcolor: '#894b9d',
+        linecolor: '#ccc',
+        circlecolor: 'white',
+        padding: 2
+      };
+
+      angular.forEach(attrs, function(value, key) {
+        options[key] = value;
+      });
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js]: `angular.merge()` maybe is better option?

File: src/invenio-search-js/invenioSearch.module.js [hunk 3]
```diff
@@ -1338,6 +1337,135 @@
...
+            var rMin = +vm.invenioSearchArgs[options.name + '.from'];
+            var rMax = +vm.invenioSearchArgs[options.name + '.to'];
+            if (!isNaN(rMin) && !isNaN(rMax)) {
+              options.selectionRange = {
+                min: rMin,
+                max: rMax
+              };
+            }
+
+            invenioSearchRangeFactory(options.histogramid, options.selectionid,
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js]: More "javascript" way :P

``` javascript
invenioSearchRangeFactory(
   options.histogramid,
   options.selectionid,
   buckets,
   options,
   changeUserSelection
)
```

File: example/index.html [hunk 4]
```diff
@@ -41,6 +43,13 @@
             <invenio-search search-endpoint="http://localhost:5000/api/records/">
               <div class="row">
                 <div class="col-md-3">
+                  <invenio-search-range
+                   template="/src/invenio-search-js/templates/range.html"
+                   histogramid="#year_hist"
```
Comment (drjova) [inline example/index.html]: can we pass all options as `JSON`? i.e.

``` html
<invenio-search-range
  template="/src/invenio-search-js/templates/range.html"
  options='{"histogramID": "#year_hist", "width": "180"}'
></invenio-search-range>
```

is pretty handy for passing options with a Jinja `config` variable

File: karma-src.conf.js [hunk 5]
```diff
@@ -51,6 +51,8 @@ module.exports = function(config) {
       'node_modules/jquery/dist/jquery.js',
       'node_modules/angular/angular.js',
       'node_modules/angular-mocks/angular-mocks.js',
+      'node_modules/d3/d3.js',
+      'node_modules/d3-tip/index.js',
```
Comment (drjova) [inline karma-src.conf.js:55]: 👍

File: src/invenio-search-js/invenioSearch.module.js [hunk 6]
```diff
@@ -1455,6 +1583,266 @@
...
+     * Renders a histogram and selection bar on the selected elements
+     *
+     * @param placement {string} - The element to contain the histogram.
+     * @param selectPlacement {string} - The element to contain the bar.
+     * @param data {Object} - The data to be displayed.
+     * @param userOptions {Object} - Options for rendering.
+     * @param userOptions.bar_color {string} - Color of the unselected bars
+     * @param userOptions.select_color {string} - Color of the selected bars
+     * @param userOptions.line_color {string} - Color of the line
+     * @param userOptions.circle_color {string} - Color of the circles on the ends of the selection
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js]: can you please add dots at the end

File: src/invenio-search-js/invenioSearch.module.js [hunk 7]
```diff
@@ -1455,6 +1583,266 @@
...
+     */
+
+    /**
+     * This callback is called when the selection changes
+     *
+     * @callback onSelection
+     * @param {number} from - Beginning of the new selection
+     * @param {number} to - End of the new selection
+     */
+
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1761]: 🙈

File: src/invenio-search-js/invenioSearch.module.js [hunk 8]
```diff
@@ -1455,6 +1583,266 @@
...
+    };
+
+    /**
+     * This callback returns the string displayed on a bar mouseover
+     *
+     * @callback tipHint
+     * @param {string} key - The key of the elements on the bar
+     * @param {number} count - The number of the elements on the bar
+     */
+
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js]: 🙈

File: test/unit/invenio-search-js/directives/searchRangeSpec.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
+/*
+ * This file is part of Invenio.
+ * Copyright (C) 2015, 2016 CERN.
```
Comment (drjova) [inline test/unit/invenio-search-js/directives/searchRangeSpec.js]: just 2016 :)

File: .jshintrc [hunk 10]
```diff
@@ -56,6 +56,7 @@
 
   "globals": {
     "angular": false,
+    "d3": false,
```
Comment (drjova) [inline .jshintrc:59]: 👍

File: src/invenio-search-js/invenioSearch.module.js (hunk 7 above)
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1761]: could you please remove this comment?

File: src/invenio-search-js/invenioSearch.module.js [hunk 11]
```diff
@@ -1338,6 +1338,138 @@
...
+          bottom: 0
+        },
+        barColor: '#2c3e50',
+        selectColor: '#894b9d',
+        lineColor: '#ccc',
+        circleColor: 'white',
+        padding: 2
+      };
+
+      var useParamsOnce = true;
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1385]: do we need that? :D

File: src/invenio-search-js/invenioSearch.module.js [hunk 12]
```diff
@@ -1338,6 +1338,138 @@
...
   }
 
+  /**
+   * @ngdoc directive
+   * @name invenioSearchRange
+   * @description
+   *    The invenioSearchRange directive
+   * @namespace invenioSearchRange
+   * @example
+   *    Usage:
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1348]: Could you please add the minimum example? as on the rest directives?

File: src/invenio-search-js/invenioSearch.module.js [hunk 13]
```diff
@@ -1455,6 +1587,275 @@
...
+        // whilst brushing is happening, this function is called
+        .on('brush', on_brushed)
+        // when finished, brushend is called
+        .on('brushend', function () {
+          var extent;
+
+          if (selected.length === 0) {
+            extent = [rangeStart, rangeEnd];
+          }
+          else {
```
Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1711]:
```
} else {
```

File: src/invenio-search-js/invenioSearch.module.js (hunk 11 above)
Comment (nikofil) [inline src/invenio-search-js/invenioSearch.module.js]: Do you mean the entire options object, or just the last line?

Comment (drjova) [inline src/invenio-search-js/invenioSearch.module.js:1385]:
``` js
var useParamsOnce = true;
```

**Pull request closed without merging**